docker/hq-compose-runserver.yml

```diff
@@ -1,34 +1,49 @@
 version: '2.3'
 
+# x-anchors are ignored by Docker, are used for DRYness here
+x-commcarehq_base: &commcarehq_base
+  build:
+    context: .
+    dockerfile: "${DOCKERFILE}"
+  environment: &environment_base
+    COMMCAREHQ_BOOTSTRAP: "yes"
+    DEPENDENT_SERVICES: "couch:5984 postgres:5432 redis:6379 elasticsearch2:9200 minio:9980 kafka:9092"
+    DOCKER_HQ_OVERLAY: "${DOCKER_HQ_OVERLAY}"
+    DOCKER_HQ_OVERLAYFS_CHMOD: "${DOCKER_HQ_OVERLAYFS_CHMOD}"
+    DOCKER_HQ_OVERLAYFS_METACOPY: "${DOCKER_HQ_OVERLAYFS_METACOPY}"
+  privileged: true  # allows mount inside container
+  depends_on:
+    - postgres
+    - couch
+    - redis
+    - elasticsearch2
+    - elasticsearch5
+    - kafka
+    - zookeeper
+    - minio
+  volumes:
+    - ..:/mnt/commcare-hq-ro:ro
```

Comment:
@joxl or @millerdev When mounting a single host directory into multiple containers, we are getting an error [here](https://github.com/dimagi/commcare-hq/blob/master/docker/run.sh#L289) at recursive chmod. This is happening only if start all those containers in parallel.  It doesn't happen if we start them one after another.

Do you recommend any best practice around mounting a single host directory over multiple containers?

Reply:
What is the error?

Reply:
@millerdev 
So when I run `DOCKER_HQ_OVERLAY=overlayfs ./scripts/docker runserver up` it tries to start `hqservice_web_1`, `hqservice_celery_1` and `hqservice_pillowtop_1` , but only 1 of these 3 will ever start, since the other 2 will usually fail with a message like below:
```
celery_1          | ln: failed to create symbolic link 'lib/overlay/node_modules/node_modules': File exists
web_1             | ln: failed to create symbolic link 'lib/overlay/node_modules/node_modules': File exists
```
(sometimes web_1 starts but pillowtop_1 fails...there's no pattern in this regard as to which services "get to succeed" it seems)

Now, when I try to start (let's suppose celery and web as above) afterwards at the same time (with `DOCKER_HQ_OVERLAY=overlayfs ./scripts/docker runserver up web` and `DOCKER_HQ_OVERLAY=overlayfs ./scripts/docker runserver up celery`), I get the below error for one of the services (in this case celery, because I started celery first...whichever one you start first will fail with the below):
```
Attaching to hqservice_celery_1
celery_1          | [run.sh] INFO: mounting /mnt/commcare-hq via overlayfs
celery_1          | [run.sh] INFO: chmod'ing commcare-hq overlay... chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_hqwebapp_management_commands_build_requirejs_py.html': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_reports_formdetails___init___py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_userreports_reports_util_py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_styleguide_management___init___py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_export_management_commands_process_skipped_pages_py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_dump_reload_management_commands_load_domain_data_py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_app_manager_app_schemas_session_schema_py.html': Operation not permitted
celery_1          | chmod: changing permissions of 'commcare-hq/coverage-report/corehq_apps_products_tasks_py.html': Operation not permitted

...

celery_1          | chmod: changing permissions of 'commcare-hq/submodules': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/README.md': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/requirements.txt': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/couchdbkit_aggregate': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/couchdbkit_aggregate/_design': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/couchdbkit_aggregate/_design/views': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/couchdbkit_aggregate/_design/views/lib': No such file or directory
celery_1          | chmod: changing permissions of 'commcare-hq/submodules/couchdbkit-aggregate/couchdbkit_aggregate/_design/views/lib/emit_array.js': No such file or directory
```
The other service will then start up correctly.

Reply:
> `ln: failed to create symbolic link 'lib/overlay/node_modules/node_modules': File exists`

This is not the recursive `chmod` failing, this is `ln` failing.  What host directory our you mounting into multiple containers, and what is its mountpoint inside those containers?

Reply:
At first glance, it seems like the `/mnt` directory inside the three containers is _the same directory_ shared across all three containers. I haven't wrapped my head around why this might be (I don't _think_ they could be sharing a common layer in the overlay). In any case, I suspect when these three containers try to spin up at the same time, they all try to perform the FS setup boilerplate on the _same OS directory_ (that's the only way I can understand why [these `ln` commands could fail](https://github.com/dimagi/commcare-hq/blob/be10a007d61d654836b2563e6cfc1feb9d4fa02c/docker/run.sh#L274-L275)). If that's indeed what's happening, then more work is needed to isolate these containers from each other, because each container expects to fully control the contents of these directories, and they will stomp on each other if they share it.